File: archive/python-legacy/polymarket/data_pipeline/multi_source_aggregator.py
"""
Multi-source weather data aggregator
Fetches from all available sources and merges into a unified dataset
"""

import logging
import pandas as pd
import numpy as np
from typing import List, Dict, Optional

from polymarket.data_pipeline.open_meteo_fetcher import OpenMeteoFetcher
from polymarket.data_pipeline.nws_fetcher import NWSFetcher
from polymarket.data_pipeline.noaa_fetcher import NOAAFetcher, MAJOR_STATIONS
from polymarket.data_pipeline.openweathermap_fetcher import OpenWeatherMapFetcher
from polymarket.data_pipeline.visual_crossing_fetcher import VisualCrossingFetcher
from polymarket.data_pipeline.weatherapi_fetcher import WeatherAPIFetcher
from polymarket.data_pipeline.tomorrow_io_fetcher import TomorrowIOFetcher

logger = logging.getLogger(__name__)

CITY_NOAA_STATIONS = {
    "NYC": "GHCND:USW00023023",
    "LA": "GHCND:USW00012918",
    "Chicago": "GHCND:USW00094846",
    "Dallas": "GHCND:USW00013060",
    "Denver": "GHCND:USW00014827",
    "Miami": "GHCND:USW00093017",
    "Boston": "GHCND:USW00024234",
}


class MultiSourceAggregator:
    """
    Aggregates weather data from multiple sources into a single unified dataset.
    Handles graceful degradation when API keys are missing.
    """

    def __init__(self):
        self.open_meteo = OpenMeteoFetcher()
        self.nws = NWSFetcher()
        self.noaa = NOAAFetcher()
        self.owm = OpenWeatherMapFetcher()
        self.vc = VisualCrossingFetcher()
        self.wapi = WeatherAPIFetcher()
        self.tio = TomorrowIOFetcher()

    def fetch_all_sources(
        self,
        location_key: str,
        start_date: str,
        end_date: str,
    ) -> Dict[str, pd.DataFrame]:
        """
        Fetch data from all available sources for a location

        Args:
            location_key: Location key (e.g., "NYC", "LA", "London")
            start_date: Start date (YYYY-MM-DD)
            end_date: End date (YYYY-MM-DD)

        Returns:
            Dictionary mapping source name to DataFrame
        """
        results = {}

        # 1. Open-Meteo (always available, no API key)
        logger.info(f"Fetching Open-Meteo data for {location_key}...")
        df = self.open_meteo.fetch_location(location_key, start_date, end_date)
        if not df.empty:
            results["open_meteo"] = df

        # 2. NWS (US only, no API key)
        logger.info(f"Fetching NWS data for {location_key}...")
        df = self.nws.fetch_location(location_key, start_date, end_date)
        if not df.empty:
            results["nws"] = df

        # 3. NOAA (US only, requires API key)
        noaa_station = CITY_NOAA_STATIONS.get(location_key)
        if noaa_station:
            logger.info(f"Fetching NOAA data for {location_key}...")
            try:
                df = self.noaa.fetch_daily_observations(
                    noaa_station, start_date, end_date
                )
                if not df.empty:
                    df = self._normalize_noaa_data(df, noaa_station)
                    results["noaa"] = df
            except Exception as e:
                logger.warning(f"NOAA fetch failed for {location_key}: {e}")

        # 4. OpenWeatherMap (requires API key)
        if self.owm.is_available():
            logger.info(f"Fetching OpenWeatherMap data for {location_key}...")
            df = self.owm.fetch_location(location_key, start_date, end_date)
            if not df.empty:
                results["openweathermap"] = df

        # 5. Visual Crossing (requires API key)
        if self.vc.is_available():
            logger.info(f"Fetching Visual Crossing data for {location_key}...")
            df = self.vc.fetch_location(location_key, start_date, end_date)
            if not df.empty:
                results["visual_crossing"] = df

        # 6. WeatherAPI (requires API key)
        if self.wapi.is_available():
            logger.info(f"Fetching WeatherAPI data for {location_key}...")
            df = self.wapi.fetch_location(location_key, start_date, end_date)
            if not df.empty:
                results["weatherapi"] = df

        # 7. Tomorrow.io (requires API key)
        if self.tio.is_available():
            logger.info(f"Fetching Tomorrow.io data for {location_key}...")
            df = self.tio.fetch_location(location_key, start_date, end_date)
            if not df.empty:
                results["tomorrow_io"] = df

        logger.info(
            f"Fetched data from {len(results)} sources for {location_key}: "
            f"{list(results.keys())}"
        )
        return results

    def _normalize_noaa_data(self, df: pd.DataFrame, station_id: str) -> pd.DataFrame:
        """Normalize raw NOAA data to standard format"""
        if df.empty:
            return df

        # NOAA returns data in long format, pivot to wide
        pivot = df.pivot_table(
            index="date",
            columns="datatype",
            values="value",
            aggfunc="first",
        ).reset_index()

        result = pd.DataFrame({
            "date": pd.to_datetime(pivot["date"]),
            "station_id": f"NOAA_{station_id}",
        })

        # NOAA temperatures are in tenths of degrees C
        if "TMAX" in pivot.columns:
            result["temperature_max"] = pivot["TMAX"] / 10.0
        if "TMIN" in pivot.columns:
            result["temperature_min"] = pivot["TMIN"] / 10.0
        if "TMAX" in pivot.columns and "TMIN" in pivot.columns:
            result["temperature_mean"] = (pivot["TMAX"] / 10.0 + pivot["TMIN"] / 10.0) / 2
        # NOAA precipitation is in tenths of mm
        if "PRCP" in pivot.columns:
            result["precipitation_total"] = pivot["PRCP"] / 10.0
        # NOAA wind is in tenths of m/s
        if "AWND" in pivot.columns:
            result["wind_speed_mean"] = pivot["AWND"] / 10.0

        return result

    def aggregate(
        self,
        location_key: str,
        start_date: str,
        end_date: str,
    ) -> pd.DataFrame:
        """
        Fetch from all sources and merge into a single dataset.
        Uses mean of all available sources for each field on each date.

        Args:
            location_key: Location key
            start_date: Start date (YYYY-MM-DD)
            end_date: End date (YYYY-MM-DD)

        Returns:
            Unified DataFrame with averaged values and source metadata
        """
        source_data = self.fetch_all_sources(location_key, start_date, end_date)

        if not source_data:
            logger.warning(f"No data from any source for {location_key}")
            return pd.DataFrame()

        weather_cols = [
            "temperature_max", "temperature_min", "temperature_mean",
            "precipitation_total", "wind_speed_mean",
        ]

        # Collect all data by date
        all_frames = []
        for source_name, df in source_data.items():
            df = df.copy()
            df["date"] = pd.to_datetime(df["date"]).dt.normalize()
            df["source"] = source_name
            all_frames.append(df[["date", "source"] + [c for c in weather_cols if c in df.columns]])

        combined = pd.concat(all_frames, ignore_index=True)

        # Average across sources for each date
        aggregated = combined.groupby("date").agg(
            temperature_max=("temperature_max", "mean"),
            temperature_min=("temperature_min", "mean"),
            temperature_mean=("temperature_mean", "mean"),
            precipitation_total=("precipitation_total", "mean"),
            wind_speed_mean=("wind_speed_mean", "mean"),
            source_count=("source", "nunique"),
            sources=("source", lambda x: ",".join(sorted(x.unique()))),
        ).reset_index()

        aggregated["station_id"] = f"MULTI_{location_key}"
        aggregated["location"] = location_key

        # Add quality score based on number of sources
        max_sources = aggregated["source_count"].max()
        aggregated["data_quality_score"] = aggregated["source_count"] / max(max_sources, 1)
        aggregated["is_validated"] = aggregated["source_count"] >= 2

        logger.info(
            f"Aggregated {len(aggregated)} days for {location_key} "
            f"from {aggregated['source_count'].max()} sources max"
        )

        return aggregated

    def aggregate_multiple(
        self,
        location_keys: List[str],
        start_date: str,
        end_date: str,
    ) -> Dict[str, pd.DataFrame]:
        """
        Aggregate data for multiple locations

        Args:
            location_keys: List of location keys
            start_date: Start date
            end_date: End date

        Returns:
            Dictionary mapping location key to aggregated DataFrame
        """
        results = {}
        for key in location_keys:
            df = self.aggregate(key, start_date, end_date)
            if not df.empty:
                results[key] = df
        return results

File: src/data_pipeline/multi_source_aggregator.rs
use std::collections::HashMap;

use chrono::NaiveDate;

use crate::data_pipeline::noaa_fetcher::NOAAFetcher;
use crate::data_pipeline::nws_fetcher::NWSFetcher;
use crate::data_pipeline::open_meteo_fetcher::OpenMeteoFetcher;
use crate::data_pipeline::openweathermap_fetcher::OpenWeatherMapFetcher;
use crate::data_pipeline::tomorrow_io_fetcher::TomorrowIOFetcher;
use crate::data_pipeline::visual_crossing_fetcher::VisualCrossingFetcher;
use crate::data_pipeline::weatherapi_fetcher::WeatherAPIFetcher;
use crate::types::{RawObservation, WeatherRecord};

#[derive(Clone)]
pub struct MultiSourceAggregator {
    pub open_meteo: OpenMeteoFetcher,
    pub nws: NWSFetcher,
    pub noaa: NOAAFetcher,
    pub owm: OpenWeatherMapFetcher,
    pub vc: VisualCrossingFetcher,
    pub wapi: WeatherAPIFetcher,
    pub tio: TomorrowIOFetcher,
}

impl MultiSourceAggregator {
    pub fn new() -> Self {
        Self {
            open_meteo: OpenMeteoFetcher::new(),
            nws: NWSFetcher::new(),
            noaa: NOAAFetcher::new(),
            owm: OpenWeatherMapFetcher::new(None),
            vc: VisualCrossingFetcher::new(None),
            wapi: WeatherAPIFetcher::new(None),
            tio: TomorrowIOFetcher::new(None),
        }
    }

    pub fn fetch_all_sources(
        &self,
        location_key: &str,
        start_date: NaiveDate,
        end_date: NaiveDate,
    ) -> HashMap<String, Vec<WeatherRecord>> {
        let mut results = HashMap::new();

        let om = self.open_meteo.fetch_location(location_key, start_date, end_date);
        if !om.is_empty() {
            results.insert("open_meteo".to_string(), om);
        }

        let nws = self.nws.fetch_location(location_key, start_date, end_date);
        if !nws.is_empty() {
            results.insert("nws".to_string(), nws);
        }

        if let Some(station) = city_noaa_stations().get(location_key) {
            let raw = self
                .noaa
                .fetch_daily_observations(station, start_date, end_date, None);
            let normalized = normalize_noaa_data(&raw, station);
            if !normalized.is_empty() {
                results.insert("noaa".to_string(), normalized);
            }
        }

        if self.owm.is_available() {
            let rows = self.owm.fetch_location(location_key, start_date, end_date);
            if !rows.is_empty() {
                results.insert("openweathermap".to_string(), rows);
            }
        }

        if self.vc.is_available() {
            let rows = self.vc.fetch_location(location_key, start_date, end_date);
            if !rows.is_empty() {
                results.insert("visual_crossing".to_string(), rows);
            }
        }

        if self.wapi.is_available() {
            let rows = self.wapi.fetch_location(location_key, start_date, end_date);
            if !rows.is_empty() {
                results.insert("weatherapi".to_string(), rows);
            }
        }

        if self.tio.is_available() {
            let rows = self.tio.fetch_location(location_key, start_date, end_date);
            if !rows.is_empty() {
                results.insert("tomorrow_io".to_string(), rows);
            }
        }

        results
    }

    pub fn aggregate(
        &self,
        location_key: &str,
        start_date: NaiveDate,
        end_date: NaiveDate,
    ) -> Vec<WeatherRecord> {
        let source_data = self.fetch_all_sources(location_key, start_date, end_date);
        if source_data.is_empty() {
            return Vec::new();
        }

        let mut date_map: HashMap<NaiveDate, Vec<(String, WeatherRecord)>> = HashMap::new();
        for (source, rows) in source_data {
            for row in rows {
                date_map
                    .entry(row.date)
                    .or_default()
                    .push((source.clone(), row));
            }
        }

        let mut dates: Vec<NaiveDate> = date_map.keys().copied().collect();
        dates.sort();

        let mut out = Vec::new();
        for date in dates {
            let rows = date_map.remove(&date).unwrap_or_default();
            let mut source_names = Vec::new();
            let mut temp_max = Vec::new();
            let mut temp_min = Vec::new();
            let mut temp_mean = Vec::new();
            let mut precip = Vec::new();
            let mut wind = Vec::new();

            for (source, row) in rows {
                source_names.push(source);
                if let Some(v) = row.temperature_max {
                    temp_max.push(v);
                }
                if let Some(v) = row.temperature_min {
                    temp_min.push(v);
                }
                if let Some(v) = row.temperature_mean {
                    temp_mean.push(v);
                }
                if let Some(v) = row.precipitation_total {
                    precip.push(v);
                }
                if let Some(v) = row.wind_speed_mean {
                    wind.push(v);
                }
            }

            source_names.sort();
            source_names.dedup();

            let mut agg = WeatherRecord::new(date, format!("MULTI_{location_key}"));
            agg.location = Some(location_key.to_string());
            agg.temperature_max = mean_opt(&temp_max);
            agg.temperature_min = mean_opt(&temp_min);
            agg.temperature_mean = mean_opt(&temp_mean);
            agg.precipitation_total = mean_opt(&precip);
            agg.wind_speed_mean = mean_opt(&wind);
            agg.source_count = Some(source_names.len() as u32);
            agg.sources = Some(source_names.join(","));
            agg.data_quality_score = Some(1.0);
            agg.is_validated = source_names.len() >= 2;
            out.push(agg);
        }

        out
    }

    pub fn aggregate_multiple(
        &self,
        location_keys: &[String],
        start_date: NaiveDate,
        end_date: NaiveDate,
    ) -> HashMap<String, Vec<WeatherRecord>> {
        let mut out = HashMap::new();

        for key in location_keys {
            let rows = self.aggregate(key, start_date, end_date);
            if !rows.is_empty() {
                out.insert(key.clone(), rows);
            }
        }

        out
    }
}

impl Default for MultiSourceAggregator {
    fn default() -> Self {
        Self::new()
    }
}

fn mean_opt(values: &[f64]) -> Option<f64> {
    if values.is_empty() {
        None
    } else {
        Some(values.iter().sum::<f64>() / values.len() as f64)
    }
}

fn normalize_noaa_data(raw: &[RawObservation], station_id: &str) -> Vec<WeatherRecord> {
    let mut map: HashMap<NaiveDate, WeatherRecord> = HashMap::new();

    for obs in raw {
        let entry = map
            .entry(obs.date)
            .or_insert_with(|| WeatherRecord::new(obs.date, format!("NOAA_{station_id}")));

        match obs.datatype.as_str() {
            "TMAX" => entry.temperature_max = Some(obs.value / 10.0),
            "TMIN" => entry.temperature_min = Some(obs.value / 10.0),
            "PRCP" => entry.precipitation_total = Some(obs.value / 10.0),
            "AWND" => entry.wind_speed_mean = Some(obs.value / 10.0),
            _ => {}
        }
    }

    let mut out: Vec<WeatherRecord> = map.into_values().collect();
    out.sort_by_key(|r| r.date);
    for row in &mut out {
        row.temperature_mean = match (row.temperature_max, row.temperature_min) {
            (Some(mx), Some(mn)) => Some((mx + mn) / 2.0),
            _ => None,
        };
    }
    out
}

fn city_noaa_stations() -> HashMap<String, String> {
    [
        ("NYC", "GHCND:USW00023023"),
        ("LA", "GHCND:USW00012918"),
        ("Chicago", "GHCND:USW00094846"),
        ("Dallas", "GHCND:USW00013060"),
        ("Denver", "GHCND:USW00014827"),
        ("Miami", "GHCND:USW00093017"),
        ("Boston", "GHCND:USW00024234"),
    ]
    .into_iter()
    .map(|(k, v)| (k.to_string(), v.to_string()))
    .collect()
}
